### PR TITLE
Handle unsupported keys and duty type mismatches (F-ssv-184)

### PR DESCRIPTION
**Pull request by julienh-ssv**
Fix for F-ssv-184
Multiple runner implementations use unchecked type assertions (`.(type)` without `ok` check) on the `CurrentDuty` field and `decidedValue`. Additionally, the P2P key conversion functions `ECDSAPrivFromInterface` and `ECDSAPubFromInterface` perform bare type assertions on peer-provided keys. If the underlying type doesn't match, this causes a panic that crashes the node.

The P2P key conversion case is particularly concerning because `ECDSAPubFromInterface` is called from the discovery path (`DiscV5Service.Node`) where it processes peer-provided public keys. If a peer presents a non-Secp256k1 key (e.g., Ed25519, which libp2p supports), the assertion panics.

Impact: Node crash, which could cause missed duties and potential slashing if the node doesn't restart promptly. The runner assertions are primarily internal consistency issues, but the P2P key conversion panics are directly exploitable by any peer presenting a non-Secp256k1 key.

Recommendation: Use checked type assertions (`value, ok := x.(Type)`) and return appropriate errors instead of panicking. Example:
```go
secpKey, ok := privkey.(*crypto.Secp256k1PrivateKey)
if !ok {
    return nil, fmt.Errorf("unsupported key type: expected Secp256k1, got %T", privkey)
}
```

Fix for https://github.com/ssvlabs/ssv-node-board/issues/762

### CODE REVIEW

**Review by nkryuchkov — DISMISSED** [review]

File: protocol/v2/ssv/runner/committee.go [hunk 1]
```diff
@@ -235,23 +235,23 @@ func (r *CommitteeRunner) ProcessConsensus(ctx context.Context, logger *zap.Logg
 	r.measurements.EndConsensus()
 	recordConsensusDuration(ctx, r.measurements.ConsensusTime(), spectypes.RoleCommittee)
 
-	duty := r.State.CurrentDuty
+	committeeDuty, err := committeeDutyFromState(r.State)
+	if err != nil {
+		return err
+	}
+	slot := committeeDuty.DutySlot()
```
Comment (iurii-ssv) [inline protocol/v2/ssv/runner/committee.go]: `slot`->`commiteeDutySlot` since `slot` can refer to many different things (message-slot, current-duty slot, estimated-current-slot, etc.).

But also, for consistency, we should either define this helper-var in every runner like that, or just inline `committeeDuty.DutySlot()` instead (TBH inlining in this case would probably be my choice)

File: protocol/v2/ssv/runner/committee.go [hunk 2]
```diff
@@ -270,12 +270,17 @@ func (r *CommitteeRunner) ProcessConsensus(ctx context.Context, logger *zap.Logg
...
-		beaconVote = decidedValue.(*spectypes.BeaconVote)
+		beaconVote *spectypes.BeaconVote
 		totalAttesterDuties,
 		totalSyncCommitteeDuties,
 		blockedAttesterDuties atomic.Uint32
 	)
 
+	beaconVote, err = beaconVoteFromEncoder(decidedValue)
+	if err != nil {
+		return err
```
Comment (iurii-ssv) [inline protocol/v2/ssv/runner/committee.go]: Lets try to wrap errors instead of `return err` whenever possible, applies to multiple places in this PR.

It's kind of "out of scope of this PR", but I don't see how else we can add wrap errors consistently if we don't do it gradually over time.

File: protocol/v2/ssv/runner/committee.go [hunk 3]
```diff
@@ -899,6 +904,10 @@ func (r *CommitteeRunner) expectedPostConsensusRootsAndBeaconObjects(ctx context
 	syncCommitteeMap = make(map[phase0.ValidatorIndex][32]byte)
 	beaconObjects = make(map[phase0.ValidatorIndex]map[[32]byte]any)
 	duty := r.State.CurrentDuty
+	committeeDuty, err := committeeDutyFromState(r.State)
+	if err != nil {
+		return nil, nil, nil, err
+	}
```
Comment (iurii-ssv) [inline protocol/v2/ssv/runner/committee.go]: We can delete this `duty := r.State.CurrentDuty` line now, and just re-use `committeeDuty` below

File: protocol/v2/ssv/runner/type_assertions.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		return nil, fmt.Errorf("duty is not a ValidatorDuty: %T", duty)
+	}
+	if validatorDuty == nil {
+		return nil, fmt.Errorf("validator duty is nil")
+	}
+
+	return validatorDuty, nil
+}
+
+func validatorDutyFromState(state *State) (*spectypes.ValidatorDuty, error) {
```
Comment (iurii-ssv) [inline protocol/v2/ssv/runner/type_assertions.go]: For consistency/simplicity, `validatorDutyFromState` should be a method on `{{VALIDTOR_DUTY_TYPE}}Runner` (eg. `ProposerRunner`) and called `currentValidatorDuty`.

We could also place this method on `baseRunner` instead to avoid duplication, but I think duplication is fine in this case.

**Review by iurii-ssv — COMMENTED** [review]
Great changes! I would ask for some updates though to keep things consistent.

File: protocol/v2/ssv/runner/type_assertions.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		return nil, fmt.Errorf("duty is not a CommitteeDuty: %T", duty)
+	}
+	if committeeDuty == nil {
+		return nil, fmt.Errorf("committee duty is nil")
+	}
+
+	return committeeDuty, nil
+}
+
+func committeeDutyFromState(state *State) (*spectypes.CommitteeDuty, error) {
```
Comment (iurii-ssv) [inline protocol/v2/ssv/runner/type_assertions.go]: For consistency/simplicity, `committeeDutyFromState` should be a method on `CommiteeRunner` and called `currentCommiteeDuty`

File: protocol/v2/ssv/runner/voluntary_exit.go [hunk 6]
```diff
@@ -223,7 +233,12 @@ func (r *VoluntaryExitRunner) executeDuty(ctx context.Context, logger *zap.Logge
 // Returns *phase0.VoluntaryExit object with current epoch and own validator index
 func (r *VoluntaryExitRunner) calculateVoluntaryExit() (*phase0.VoluntaryExit, error) {
 	epoch := r.NetworkConfig.EstimatedEpochAtSlot(r.State.CurrentDuty.DutySlot())
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/voluntary_exit.go]: Could this still panic on nil `CurrentDuty`? Seems that `r.State.CurrentDuty.DutySlot()` runs before `validatorDutyFromState` can guard against a nil duty, which defeats the safety goal of this PR. Maybe swap the order — fetch `duty` first and use `duty.DutySlot()` for the epoch calc.

File: protocol/v2/ssv/runner/validator_registration.go [hunk 7]
```diff
@@ -178,6 +183,11 @@ func (r *ValidatorRegistrationRunner) executeDuty(ctx context.Context, logger *z
 	// Reuse the existing span instead of generating new one to keep tracing-data lightweight.
 	span := trace.SpanFromContext(ctx)
 
+	validatorDuty, err := validatorDutyFromDuty(duty)
+	if err != nil {
+		return err
+	}
+
 	vr, err := r.buildValidatorRegistration(duty.DutySlot())
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/validator_registration.go]: Any reason we're not using `validatorDuty.DutySlot()` here? Every other call site in this function switched to `validatorDuty`, so `duty.DutySlot()` reads as a leftover. Not a bug (the interface exposes `DutySlot`), just consistency.

File: protocol/v2/ssv/runner/committee.go [hunk 8]
```diff
@@ -899,6 +904,10 @@ func (r *CommitteeRunner) expectedPostConsensusRootsAndBeaconObjects(ctx context
 	syncCommitteeMap = make(map[phase0.ValidatorIndex][32]byte)
 	beaconObjects = make(map[phase0.ValidatorIndex]map[[32]byte]any)
 	duty := r.State.CurrentDuty
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/committee.go]: Worth considering dropping this `duty` local entirely? It's also used at line 917 (`slot := duty.DutySlot()`); swapping that to `committeeDuty.DutySlot()` would let this line go away, which also addresses the earlier note about removing it.

**Review by iurii-ssv — DISMISSED** [review]

**Review by nkryuchkov — DISMISSED** [review]

File: network/commons/keys_test.go
Comment (oleg-ssvlabs) [inline network/commons/keys_test.go:1]: unit tests do not cover `nil` key inputs cases

File: protocol/v2/ssv/runner/aggregator.go [hunk 9]
```diff
@@ -132,7 +137,10 @@ func (r *AggregatorRunner) ProcessPreConsensus(ctx context.Context, logger *zap.
 		return fmt.Errorf("got pre-consensus quorum but it has invalid signatures: %w", err)
 	}
 
-	duty := r.State.CurrentDuty.(*spectypes.ValidatorDuty)
+	duty, err := r.currentValidatorDuty()
+	if err != nil {
+		return err
```
Comment (oleg-ssvlabs) [inline protocol/v2/ssv/runner/aggregator.go]: all returned errors from this methods are wrapped, so to be consistent, it is better to wrap this error as well. It relates to other changes in the file as well

**Review by nkryuchkov — DISMISSED** [review]

File: protocol/v2/ssv/runner/proposer.go [hunk 10]
```diff
@@ -137,7 +142,10 @@ func (r *ProposerRunner) ProcessPreConsensus(ctx context.Context, logger *zap.Lo
 		return fmt.Errorf("got pre-consensus quorum but it has invalid signatures: %w", err)
 	}
 
-	duty := r.State.CurrentDuty.(*spectypes.ValidatorDuty)
+	duty, err := r.currentValidatorDuty()
+	if err != nil {
```
Comment (oleg-ssvlabs) [inline protocol/v2/ssv/runner/proposer.go:146]: Better to wrap errors in this file as well (where it needs to be wrapped)

**Review by iurii-ssv — DISMISSED** [review]

File: protocol/v2/ssv/runner/sync_committee_contribution.go [hunk 11]
```diff
@@ -155,7 +160,10 @@ func (r *SyncCommitteeAggregatorRunner) ProcessPreConsensus(ctx context.Context,
 		return nil
 	}
 
-	duty := r.State.CurrentDuty.(*spectypes.ValidatorDuty)
+	duty, err := r.currentValidatorDuty()
+	if err != nil {
```
Comment (oleg-ssvlabs) [inline protocol/v2/ssv/runner/sync_committee_contribution.go:164]: ditto

File: protocol/v2/ssv/runner/type_assertions_test.go
Comment (oleg-ssvlabs) [inline protocol/v2/ssv/runner/type_assertions_test.go:1]: `type_assertions_test.go` covers happy paths and wrong-type errors, but not the `nil` and typed-nil scenarios that the new helpers explicitly add. Please add cases for `nil` `Duty`/`Encoder`, typed-nil `validator`/`committee` duties, typed-nil `decided` values, and the `nil` `runner/state/current-duty` scenarios for `currentCommitteeDuty`.

**Review by oleg-ssvlabs — CHANGES_REQUESTED** [review]
gj, left few comments

File: protocol/v2/ssv/runner/committee.go
Comment (iurii-ssv) [inline protocol/v2/ssv/runner/committee.go]: We should use `committeeDutyFromDuty` func instead of this block of code here in `CommitteeRunner.StartNewDuty` to match what we do for Validator-Runners

File: protocol/v2/ssv/runner/type_assertions.go [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+func (b *BaseRunner) currentValidatorDuty() (*spectypes.ValidatorDuty, error) {
+	if b == nil {
+		return nil, fmt.Errorf("runner is nil")
+	}
+	if b.State == nil {
+		return nil, fmt.Errorf("runner state is nil")
+	}
+	if b.State.CurrentDuty == nil {
+		return nil, fmt.Errorf("current duty is nil")
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/type_assertions.go:34]: Worth considering — many `r.State.CurrentDuty.DutySlot()` direct accesses still remain in `runner_validations.go`, `proposer.go` (`expectedPreConsensusRootsAndDomain` etc.), `sync_committee_contribution.go`, `committee.go`, `aggregator.go` and `validator_registration.go` that bypass these guards. And `runner.go:515` already documents `// assume CurrentDuty is not nil if state is not nil`, so the codebase already relies on that invariant.

I fear the in-between state (guards here, raw access elsewhere) is inconsistent — either extend the helper to those sites, or drop the `State == nil` / `CurrentDuty == nil` checks here since the invariant is already assumed upstream.

File: protocol/v2/ssv/runner/committee.go [hunk 13]
```diff
@@ -235,23 +235,23 @@ func (r *CommitteeRunner) ProcessConsensus(ctx context.Context, logger *zap.Logg
 	r.measurements.EndConsensus()
 	recordConsensusDuration(ctx, r.measurements.ConsensusTime(), spectypes.RoleCommittee)
 
-	duty := r.State.CurrentDuty
+	committeeDuty, err := r.currentCommitteeDuty()
+	if err != nil {
+		return fmt.Errorf("current committee duty: %w", err)
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/committee.go:240]: Could this be simplified? The helper already wraps the inner error with `current duty: %w`, so this ends up producing `current committee duty: current duty: duty is not a CommitteeDuty: ...`. Maybe drop one of the wraps — a bare `return err` here (or removing the inner wrap inside `currentCommitteeDuty`) would read cleaner. Same pattern applies at `expectedPostConsensusRootsAndBeaconObjects` (line ~907) and `voluntary_exit.go` `calculateVoluntaryExit`.

File: protocol/v2/ssv/runner/committee.go [hunk 14]
```diff
@@ -270,12 +270,17 @@ func (r *CommitteeRunner) ProcessConsensus(ctx context.Context, logger *zap.Logg
 		signaturesCh = make(chan *spectypes.PartialSignatureMessage)
 		dutiesCh     = make(chan *spectypes.ValidatorDuty)
 
-		beaconVote = decidedValue.(*spectypes.BeaconVote)
+		beaconVote *spectypes.BeaconVote
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/committee.go]: Might be worth pulling `beaconVote` out of this var block — it's a forward declaration that's immediately overwritten a few lines below at L279. A standalone `beaconVote, err := beaconVoteFromEncoder(decidedValue)` would match the pattern the other runners use and read cleaner.

File: protocol/v2/ssv/runner/voluntary_exit.go [hunk 15]
```diff
@@ -159,6 +164,11 @@ func (r *VoluntaryExitRunner) executeDuty(ctx context.Context, logger *zap.Logge
 	// Reuse the existing span instead of generating new one to keep tracing-data lightweight.
 	span := trace.SpanFromContext(ctx)
 
+	validatorDuty, err := validatorDutyFromDuty(duty)
+	if err != nil {
+		return err
+	}
+
 	voluntaryExit, err := r.calculateVoluntaryExit()
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/voluntary_exit.go]: Seems that `validatorDuty` here comes from the function argument while `calculateVoluntaryExit` independently re-fetches via `r.currentValidatorDuty()` (reading from state). They should match today, but having two sources of truth for the same duty in one function feels inconsistent. Maybe pass `validatorDuty` into `calculateVoluntaryExit` so there's only one path.

File: protocol/v2/ssv/runner/type_assertions_test.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+	got, err := (&BaseRunner{State: &State{CurrentDuty: committeeDuty}}).currentCommitteeDuty()
+	require.NoError(t, err)
+	require.Same(t, committeeDuty, got)
+
+	_, err = (&BaseRunner{State: &State{CurrentDuty: &spectypes.ValidatorDuty{}}}).currentCommitteeDuty()
+	require.ErrorContains(t, err, "current duty: duty is not a CommitteeDuty")
+}
+
+func TestCalculateVoluntaryExitHandlesMissingCurrentDuty(t *testing.T) {
```
Comment (momosh-ssv) [inline protocol/v2/ssv/runner/type_assertions_test.go]: Couple of nits on this file:

- `TestCalculateVoluntaryExitHandlesMissingCurrentDuty` is really a `VoluntaryExitRunner` test — might fit better in a `voluntary_exit_test.go` for discoverability.
- `TestValidatorDutyFromState` / `TestCommitteeDutyFromState` above exercise `currentValidatorDuty` / `currentCommitteeDuty` — the `FromState` naming no longer matches what's under test.

File: network/commons/keys_test.go
Comment (julienh-ssv) [reply network/commons/keys_test.go:1]: Added nil-interface and typed-nil coverage for both `ECDSAPrivFromInterface` and `ECDSAPubFromInterface`.

File: protocol/v2/ssv/runner/aggregator.go (hunk 9 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/aggregator.go]: Wrapped the helper-returned errors in `aggregator.go` to match the rest of the file.

File: protocol/v2/ssv/runner/proposer.go (hunk 10 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/proposer.go:146]: Wrapped the helper-returned errors in `proposer.go` for consistency with the surrounding flow.

File: protocol/v2/ssv/runner/sync_committee_contribution.go (hunk 11 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/sync_committee_contribution.go:164]: Applied the same error-wrapping cleanup in `sync_committee_contribution.go`.

File: protocol/v2/ssv/runner/type_assertions_test.go
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/type_assertions_test.go:1]: Expanded the helper tests for nil, typed-nil, and runner/state/current-duty edge cases, and added the `VoluntaryExitRunner` regression in `voluntary_exit_test.go`.

File: protocol/v2/ssv/runner/committee.go
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/committee.go]: Switched `CommitteeRunner.StartNewDuty` to use `committeeDutyFromDuty`.

File: protocol/v2/ssv/runner/type_assertions.go (hunk 12 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/type_assertions.go:34]: Added `currentDutySlot()` and replaced the remaining raw `CurrentDuty.DutySlot()` reads across the runner package so the guard usage is consistent.

File: protocol/v2/ssv/runner/committee.go (hunk 13 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/committee.go:240]: Removed the double wrapping by simplifying the helper internals and keeping the contextual wrap at the call sites.

File: protocol/v2/ssv/runner/committee.go (hunk 14 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/committee.go]: Pulled `beaconVote` into a standalone assignment instead of the forward declaration in the var block.

File: protocol/v2/ssv/runner/voluntary_exit.go (hunk 15 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/voluntary_exit.go]: Changed `calculateVoluntaryExit` to take the already-validated `validatorDuty`, so there is only one duty source in that flow.

File: protocol/v2/ssv/runner/type_assertions_test.go (hunk 16 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/type_assertions_test.go]: Moved the `VoluntaryExitRunner`-specific regression into `voluntary_exit_test.go` and renamed the current-duty helper tests to match what they cover.

File: protocol/v2/ssv/runner/voluntary_exit.go [hunk 17]
```diff
@@ -221,9 +236,13 @@ func (r *VoluntaryExitRunner) executeDuty(ctx context.Context, logger *zap.Logge
...
-func (r *VoluntaryExitRunner) calculateVoluntaryExit() (*phase0.VoluntaryExit, error) {
-	epoch := r.NetworkConfig.EstimatedEpochAtSlot(r.State.CurrentDuty.DutySlot())
-	validatorIndex := r.State.CurrentDuty.(*spectypes.ValidatorDuty).ValidatorIndex
+func (r *VoluntaryExitRunner) calculateVoluntaryExit(duty *spectypes.ValidatorDuty) (*phase0.VoluntaryExit, error) {
+	if duty == nil {
+		return nil, fmt.Errorf("validator duty is nil")
+	}
+	epoch := r.NetworkConfig.EstimatedEpochAtSlot(duty.DutySlot())
+
+	validatorIndex := duty.ValidatorIndex
```
Comment (iurii-ssv) [inline protocol/v2/ssv/runner/voluntary_exit.go]: The formatting here is kinda hard to read, can we just inline these vars ?

**Review by iurii-ssv — DISMISSED** [review]

File: protocol/v2/ssv/runner/voluntary_exit.go (hunk 17 above)
Comment (julienh-ssv) [reply protocol/v2/ssv/runner/voluntary_exit.go]: Inlined the temporary fields in `calculateVoluntaryExit` to make the return block easier to scan.

**Review by iurii-ssv — APPROVED** [review]

**Review by oleg-ssvlabs — APPROVED** [review]
🥇

**Review by ljuba-ssv — APPROVED** [review]
